### DISCUSSION TITLE
Using with neodev

### DISCUSSION THREAD

**Discussion by s1341** [VonHeikemen/lsp-zero.nvim#74] [General]
Hi,

I'm attempting to use `lsp-zero` with the `neodev` plugin (https://github.com/folke/neodev.nvim), but can't get completion/lsp functionality for the `vim` global object or plugins to work.

I know `lsp-zero` has the `nvim_workspace` function, but it lacks some features (like api function documentation).
 
Any hints on how to get these plugins to play nicely with each other?

Thanks,
s1341

**Comment by VonHeikemen** [comment]
I tried this and it worked.

```lua
require('neodev').setup({})

local lsp = require('lsp-zero')

lsp.preset('recommended')
lsp.setup()
```

**Comment by s1341** [comment]
My init.lua is in ~/.config/nvim/. Does that make a diffference?

**Comment by VonHeikemen** [comment]
I don't think so. That's the right folder.

**Comment by s1341** [comment]
Ok I figured it out. `lua-language-server` wasn't running because I'm on nixos and an interpreter patch was needed.

**Comment by s1341** [comment]
For those interested, the problem was that `mason.nvim` downloads `lua-language-server` as a binary, and that binary can't run as is on nixos. See https://github.com/williamboman/mason.nvim/issues/428#issuecomment-1357192515 for a fix.

**Comment by ricardovg4** [comment]
Hi, it works for me (neodev gives the completion) but it disables the command LspZeroFormat:

E5108: Error executing lua: Command not found: LspZeroFormat
stack traceback:
        [C]: at 0x56474ff97860

**Reply by VonHeikemen** [reply]
what about the LSP keybindings? do they work?

**Reply by ricardovg4** [reply]
gq to format a single line is working. However, LspZeroFormat disappears, and also the diagnostics with gl seem to have vanished. My config file for neodev.lua is:

require("neodev").setup({
})

local lspconfig = require('lspconfig')

lspconfig.sumneko_lua.setup({
settings = {
Lua = {
completion = {
callSnippet = "Replace"
},
workspace = {
checkThirdParty = false,
}
}
}
})

I also tried doing the setup on neodev.lua like this, but then completions on neodevs part don't work:

require("neodev").setup({
})

local lspconfig = require('lsp-zero')
lspconfig.use('sumneko_lua', {
settings = {
Lua = {
completion = {
callSnippet = "Replace"
},
workspace = {
checkThirdParty = false,
}
}
}
})

**Reply by ricardovg4** [reply]
It seems that placing the config of neodev.lua inside the config for lspzero.lua and before lsp.setup() has corrected the issue. Thank you for your last comment.

**Comment by vishalbalaji** [comment]
I actually have kind of the opposite issue. I setup sumneko using lsp-zero along with custom settings, for example `settings.Lua.diagnostics.workspaceDelay = -1` to disable workspace diagnostics. However, running `neodev.setup()` overwrites this configuration, regardless of where I call the function.

I already raised an issue on the neodev repo(https://github.com/folke/neodev.nvim/issues/97) and tested with manual `nvim-lspconfig` configuration on a minimal config, where this does not seem to be an issue. It seems `neodev` needs to run before the lsp servers get setup, which might be a potential cause of this issue. Is there a way to run pre-setup hooks for `lsp-zero`?

**Reply by VonHeikemen** [reply]
There are no pre-setup hooks. If you want to run something "pre-setup", just do it before calling `lsp.setup()`.

**Reply by vishalbalaji** [reply]
I feel stupid. Turned out that there was just a duplicate `neodev.setup` call after `lsp.setup` that I missed. Thanks for replying and sorry for wasting your time :pray:.
